client/src/modules/reports/generate/dataKit/dataKit.config.js

```diff
@@ -0,0 +1,133 @@
+angular.module('bhima.controllers')
+  .controller('dataKitController', DataKitConfigController);
+
+DataKitConfigController.$inject = [
+  '$sce', 'NotifyService', 'BaseReportService', 'AppCache', 'reportData', '$state', 'SurveyFormService',
+  'ChoicesListManagementService',
+];
+
+
+function DataKitConfigController($sce, Notify, SavedReports, AppCache, reportData, $state, SurveyForm,
+  ChoicesList) {
+  const vm = this;
+  const cache = new AppCache('data_kit');
+  const reportUrl = '/data_kit/report';
+  vm.reportDetails = {};
+  vm.reportDetails.searchDateFrom = {};
+  vm.reportDetails.searchDateTo = {};
+  vm.reportDetails.multipleChoice = {};
+  vm.reportDetails.loggedChanges = {};
+  vm.onSelectList = onSelectList;
+  vm.onSelectMultiple = onSelectMultiple;
+  vm.onClear = onClear;
+
+  vm.previewGenerated = false;
+  vm.onSelectSurveyForm = onSelectSurveyForm;
+  checkCachedConfiguration();
+
+  function onSelectSurveyForm(collector) {
+    vm.reportDetails = {};
+    vm.reportDetails.searchDateFrom = {};
+    vm.reportDetails.searchDateTo = {};
+    vm.reportDetails.multipleChoice = {};
+    vm.reportDetails.loggedChanges = {};
+
+    vm.reportDetails.data_collector_id = collector.id;
+    delete cache.reportDetails;
+
+    SurveyForm.read(null, { data_collector_management_id : collector.id })
+      .then(surveyElements => {
+        vm.formItems = surveyElements;
+        return ChoicesList.read();
+      })
+      .then(choicesLists => {
+        vm.choicesLists = choicesLists;
+      })
+      .catch(Notify.handleError);
+  }
+
+  function onSelectList(list, value) {
+    vm.reportDetails[value] = list.id;
+    vm.reportDetails.loggedChanges[value] = list.label;
+  }
+
+  function onSelectMultiple(lists, value) {
+    vm.reportDetails.multipleChoice[value] = lists;
+  }
+
+  function onClear(value) {
+    delete vm.reportDetails[value];
+    delete vm.reportDetails.loggedChanges[value];
+  }
+
+  vm.preview = function preview(form) {
+    if (form.$invalid) { return; }
+    cache.reportDetails = angular.copy(vm.reportDetails);
+
+    if (vm.reportDetails.multipleChoice) {
+      const multipleChoiceLength = Object.keys(vm.reportDetails.multipleChoice).length;
+
+      if (multipleChoiceLength) {
+        Object.keys(vm.reportDetails.multipleChoice).forEach((key) => {
+          if (vm.reportDetails.multipleChoice[key].length) {
+            for (let i = 0; i < vm.reportDetails.multipleChoice[key].length; i++) {
+              vm.choicesLists.forEach(list => {
+                if (list.id === vm.reportDetails.multipleChoice[key][i]) {
```

Comment:
So .. it looks like you are taking a list of ids and turning them into a list of `label`s.  So, you are trying to do:
```js
var ids = [3, 3, 5, 7, 1]
// do something
console.log(ids) // => ['formA', 'formA', 'formZ', 'formX', 'formT']
```

Let's try to clean this up a bit.  Here is what I think the basic structure could look like:
```js
angular.forEach(vm.reportDetails.mulitpleChoice, (values, key) => {
  // values is an array.
  values.forEach(value => {
    vm.choicesLists.forEach(list => {
      if (list.id === value) {
         // replace value with list.label
      }
    });
  });
});
```

Let's see if we can go one step further:
```js
// create a way to map id -> label for the choice list
let choiceListMap = vm.choicesList.reduce((map, item) => {
  map[item.id] = item.label;
}, {});

angular.forEach(vm.reportDetails.mulitpleChoice, (values, key) => {
  // unfortunately, angular.map() doesn't exist for object.  We have to use an ugly forEach
  vm.multipleChoice[key] = values.map(value => choiceListMap[value]);
});
```

I think that should do it.

First, we create an object to make choiceList ids map to their labels.  This takes up memory but removes the `choiceLists.forEach()` loop.

Then, we use `angular.forEach()` to loop through the object.  This removes all the `Object.keys()` calls and checks for `.length`.  

Try it out and see if I've missed anything.